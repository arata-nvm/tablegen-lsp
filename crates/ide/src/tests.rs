use std::{collections::HashMap, path::Path, sync::Arc};

use crate::{
    db::{SourceDatabase, SourceDatabaseStorage},
    eval::EvalDatabaseStorage,
    file_system::{self, FileId, FilePath, FileSet, FileSystem},
};

const DEFAULT_FILE_PATH: &str = "/main.td";

pub fn single_file(fixture: &str) -> (TestDatabase, Fixture) {
    let mut f = Fixture::single_file(fixture);
    let db = TestDatabase::new(&mut f);
    (db, f)
}

pub fn multiple_files(fixture: &str) -> (TestDatabase, Fixture) {
    let mut f = Fixture::multiple_files(fixture);
    let db = TestDatabase::new(&mut f);
    (db, f)
}

#[salsa::database(SourceDatabaseStorage, EvalDatabaseStorage)]
#[derive(Default)]
pub struct TestDatabase {
    storage: salsa::Storage<Self>,
}

impl salsa::Database for TestDatabase {}

impl TestDatabase {
    fn new(f: &mut Fixture) -> Self {
        let mut db = Self::default();
        for (file_id, content) in f.files() {
            db.set_file_content(file_id, Arc::from(content));
        }

        let source_root = file_system::collect_sources(&mut db, f, f.root_file());
        db.set_source_root(Arc::new(source_root));

        db
    }
}

#[derive(Debug, Default)]
pub struct Fixture {
    file_contents: HashMap<FilePath, String>,
    file_ids: Vec<FileId>,

    file_set: FileSet,
    next_file_id: u32,
}

impl Fixture {
    fn single_file(fixture: &str) -> Self {
        let mut this = Self::default();
        this.insert_file(
            FilePath::from(Path::new(DEFAULT_FILE_PATH)),
            fixture.to_string(),
        );
        this
    }

    fn multiple_files(fixture: &str) -> Self {
        let mut this = Self::default();

        let mut lines = fixture
            .lines()
            .skip_while(|line| !line.starts_with("; "))
            .peekable();
        while let Some(line) = lines.next() {
            let path_str = line
                .strip_prefix("; ")
                .expect("expected header line to start with '; '");
            let path = FilePath::from(Path::new(path_str));

            let mut content = String::new();
            while let Some(line) = lines.peek() {
                if line.starts_with("; ") {
                    break;
                }

                let line = lines.next().expect("line must be present");
                content.push_str(line);
                content.push_str("\n");
            }

            this.insert_file(path, content);
        }

        this
    }

    fn files(&self) -> Vec<(FileId, &str)> {
        self.file_contents
            .iter()
            .map(|(path, content)| {
                (
                    self.file_set
                        .file_for_path(path)
                        .expect("file id must exist"),
                    content.as_str(),
                )
            })
            .collect()
    }

    fn root_file(&self) -> FileId {
        self.file_ids
            .first()
            .expect("at least one file must be exist")
            .clone()
    }

    fn insert_file(&mut self, path: FilePath, content: String) {
        let res = self.file_contents.insert(path.clone(), content);
        assert!(res.is_none(), "duplicate file path");

        let file_id = self.alloc_file_id();
        self.file_set.insert(file_id, path);
        self.file_ids.push(file_id);
    }

    fn alloc_file_id(&mut self) -> FileId {
        let file_id = FileId(self.next_file_id);
        self.next_file_id += 1;
        file_id
    }
}

impl FileSystem for Fixture {
    fn assign_or_get_file_id(&mut self, path: FilePath) -> FileId {
        match self.file_set.file_for_path(&path) {
            Some(file_id) => file_id,
            None => {
                let file_id = self.alloc_file_id();
                self.file_set.insert(file_id, path);
                file_id
            }
        }
    }

    fn path_for_file(&self, file_id: &FileId) -> Option<FilePath> {
        self.file_set.path_for_file(file_id)
    }

    fn read_content(&self, file_path: &FilePath) -> Option<String> {
        self.file_contents.get(file_path).cloned()
    }
}
