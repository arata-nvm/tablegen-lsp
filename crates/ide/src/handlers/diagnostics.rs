use std::collections::HashMap;

use crate::{
    eval::EvalDatabase,
    file_system::{FileId, FileRange},
};

pub fn exec(db: &dyn EvalDatabase) -> HashMap<FileId, Vec<Diagnostic>> {
    let mut diagnostic_list = Vec::new();

    let source_root = db.source_root();
    let parse = db.parse(source_root.root());
    diagnostic_list.extend(parse.errors().iter().map(|err| {
        Diagnostic::new(
            FileRange::new(source_root.root(), err.range),
            err.message.to_string(),
        )
    }));

    let evaluation = db.eval();
    diagnostic_list.extend(evaluation.diagnostics().iter().cloned());

    let mut diagnostic_map = HashMap::new();
    for diagnostic in diagnostic_list {
        let file_id = diagnostic.range.file;
        let diagnostics = diagnostic_map.entry(file_id).or_insert_with(Vec::new);
        diagnostics.push(diagnostic);
    }
    diagnostic_map
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct Diagnostic {
    pub range: FileRange,
    pub message: String,
}

impl Diagnostic {
    pub fn new(range: FileRange, message: impl Into<String>) -> Self {
        Self {
            range,
            message: message.into(),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::tests;

    #[test]
    fn syntax() {
        let (db, _) = tests::single_file("clas");
        let diags = super::exec(&db);
        insta::assert_debug_snapshot!(diags);
    }

    #[test]
    fn eval() {
        let (db, _) = tests::single_file(r#"include "not_exist.td""#);
        let diags = super::exec(&db);
        insta::assert_debug_snapshot!(diags);
    }
}
