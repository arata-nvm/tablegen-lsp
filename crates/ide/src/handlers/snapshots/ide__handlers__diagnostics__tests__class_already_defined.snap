---
source: crates/ide/src/handlers/diagnostics.rs
expression: "check(\"class Foo; class Foo { int b; }\")"
---
[]
